### PR TITLE
[2/3] Datasetchatenv

### PR DESCRIPTION
**Pull request by Darktex**
This turned out to be far less trivial than I had anticipated. What do people think of this API?

### CODE REVIEW

File: forge/rl/environments/dataset.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+            The extracted content as a string
+        """
+        content = row[self.column]
+        if self.content_template:
+            # Allow template formatting with row data
+            return self.content_template.format(**row)
+        return str(content)
+
+
+class DatasetChatEnvironment(ChatEnvironment):
```
Comment (pbontrager) [inline forge/rl/environments/dataset.py:81]: I don't understand why there needs to be a separate DatasetChatEnv and ChatEnv. Don't we just want a single chat env that wraps a dataloader and we can add tool support down the line?

File: forge/rl/environments/base.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+# LICENSE file in the root directory of this source tree.
+from __future__ import annotations
+
+import abc
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, Optional, Union
+
+
+@dataclass
+class Observation:
```
Comment (pbontrager) [inline forge/rl/environments/base.py]: If this is a standin for the Message type (I think @joecummings  proposed OAI message with extra metadata?) then I'm fine with this, but I think we should put in the comments a todo or warning so we remember to remove these.

File: forge/rl/environments/base.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                 for transforms, logging, evaluation, etc.
+    """
+
+    done: bool = False
+    reward: Optional[Union[bool, int, float]] = None
+    metadata: Dict[str, Any] = field(default_factory=dict)
+
+
+@dataclass
+class Action:
```
Comment (pbontrager) [inline forge/rl/environments/base.py]: This should be able to be the same message type, not separate from the observation

File: forge/rl/environments/base.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    - May contain information not exposed in observations
+
+    Args:
+        metadata: Additional state information that may be useful for debugging or analysis
+    """
+
+    metadata: Dict[str, Any] = field(default_factory=dict)
+
+
+class Transform(abc.ABC):
```
Comment (pbontrager) [inline forge/rl/environments/base.py]: This is the tokenizer right? In torchtune we had tokenizers as a special case of transforms but that ended up causing a lot more confusion than what it was worth. I think we should just use the tokenizer API and assume they can be multimodal. Unless this is transforming the raw output to the OAI format? I think for the latter case, we should just require that the env writer handles that themselves. For datasets we already handle this for you

File: forge/rl/environments/base.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        """Take a step in the environment and return an observation."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def state(self) -> State:
+        """Get the current state of the environment."""
+        pass
+
+    def _apply_transform(self, observation: Observation) -> Observation:
```
Comment (pbontrager) [inline forge/rl/environments/base.py]: I think we should be "dumb" here and just make the user responsible for this. We should just enforce the output type for step and input type for action and let them implement this without a standard pattern. I think it's faster and less error prone to start simpler and dumber and add more later.

File: forge/rl/environments/chat.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+from forge.rl.environments.base import Action, Environment, Observation, State
+
+
+class Message(TypedDict):
+    role: str
+    content: str
+
+
+@dataclass
+class ChatAction(Action):
```
Comment (pbontrager) [inline forge/rl/environments/chat.py]: I think these types can't be env specific right? They will be passed to and from the policy model which will always expect the same types no matter what env is used. Should we not make these data classes and just implement one version for now?

File: forge/rl/environments/chat.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+            )
+            self._state.history_tokens.append(system_tokens)
+
+    def reset(self) -> ChatObservation:
+        """Reset the environment to initial state.
+
+        Returns:
+            ChatObservation: Initial observation with system prompt (if any)
+        """
+        self._state.history_messages = []
```
Comment (pbontrager) [inline forge/rl/environments/chat.py]: I think the state data class should handle the full trajectory so you can encapsulate all of this logic inside of state and keep things simpler here.

File: forge/rl/environments/chat.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    ]
+    tokens = tensor([1, 2, 3, 4, 5, ...])  # tokenized entire conversation
+    """
+
+    messages: List[Message] = field(default_factory=list)
+    tokens: torch.Tensor = field(default_factory=lambda: torch.tensor([]))
+    # Inherited fields from Observation ABC: reward, done, metadata
+
+
+class ChatEnvironment(Environment):
```
Comment (pbontrager) [inline forge/rl/environments/chat.py]: I think right now you've actually implemented this as a dataset and the DatasetChatEnv as the env? If you look at torchtune sft dataset it already handles appending system prompts etc. I think ChatEnv should just wrap stateful dataloader and take in a dataset (might want to mock it until sft lands) and apply the tokenizer.

The question I have for you is if you think a ChatEnv should return a batch of data and then unbatch the data when we send to the policy or to always have a bs of 1 but have to spin up a lot more monarch processes (which I don't think cost anything on cpu)

File: forge/protocols/tokenizer.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (ebsmothers) [inline forge/protocols/tokenizer.py]: This is from the previous PR, right? Can you merge main so we have the minimal set of changes in this one?

File: .gitignore [hunk 10]
```diff
@@ -185,5 +185,7 @@ cover/
 # VSCode
 .vscode/
 
+PREFERENCES.md
```
Comment (ebsmothers) [inline .gitignore:188]: What's this about?

File: forge/rl/environments/dataset.py [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+        Args:
+            row: A single row from the dataset
+
+        Returns:
+            The extracted content as a string
+        """
+        content = row[self.column]
+        if self.content_template:
+            # Allow template formatting with row data
+            return self.content_template.format(**row)
```
Comment (ebsmothers) [inline forge/rl/environments/dataset.py:77]: How should I think about content_template here? Just an f-string to format a single message?

File: forge/rl/environments/base.py [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        Examples:
+            >>> target = MyObservation(done=False, reward=None)
+            >>> source = DifferentObservation(done=True, reward=1.0, extra="ignored")
+            >>> new_obs = env._convert_observation_type(target, source)
+            >>> print(type(new_obs), new_obs.done, new_obs.reward)
+            <class 'MyObservation'> True 1.0
+        """
+        new_obs = type(target_obs)()
+
+        for attr_name, attr_value in vars(target_obs).items():
```
Comment (ebsmothers) [inline forge/rl/environments/base.py]: noob q: what's the difference between this and just doing a copy here?

File: forge/rl/environments/base.py (hunk 4 above)
Comment (Darktex) [reply forge/rl/environments/base.py]: That's not the tokenizer, this is the env transforms like we had in the notebook. For example, you can do rewards here

File: forge/rl/environments/chat.py (hunk 6 above)
Comment (Darktex) [reply forge/rl/environments/chat.py]: I went back and forth on this. Maybe having a single standardized class for these is the way to go. The thing is, depending on the env people may want to dump very different things in there, so I'm not sure how to handle that

File: forge/rl/environments/chat.py (hunk 8 above)
Comment (Darktex) [reply forge/rl/environments/chat.py]: I would just do batch size 1. We can introduced batched envs as a generalization later on, but I don't see a need to have them as a p0 exactly for the reason you are saying. Where we care about batching (eg in reward computation and in VLLM generation) we have batching anyway, so this is not costing us anything

File: forge/protocols/tokenizer.py (hunk 9 above)
Comment (Darktex) [reply forge/protocols/tokenizer.py]: yea

File: .gitignore (hunk 10 above)
Comment (Darktex) [reply .gitignore:188]: Ehm trying to give Devmate my code style preferences but it doesn't remember them so I had to write something to reference :D

File: forge/rl/environments/chat.py (hunk 7 above)
Comment (Darktex) [reply forge/rl/environments/chat.py]: It is handling the full trajectory though, right? We have the whole history and the trajectory as tokens. Or did you mean smt else?

File: forge/rl/environments/dataset.py (hunk 11 above)
Comment (Darktex) [reply forge/rl/environments/dataset.py:77]: Yup. This is useful if you want the user turn to be a combination of multiple columns (it would happen in Alpaca where you need to handle a potentially missing input column)

File: forge/rl/environments/base.py (hunk 12 above)
Comment (Darktex) [reply forge/rl/environments/base.py]: Hmmm, we could do a copy of the target obs and go back and fill the values from source and it would be the same, yeah.

Maybe Philip is right that we should avoid this problem altogether and have a single Observation/Action type with a structure that allows you to add whatever you want. I'd like to point out that this path is starting to look quite similar to having a TensorDict so it feels like we are rediscovering TorchRL...

File: forge/rl/environments/dataset.py (hunk 1 above)
Comment (Darktex) [reply forge/rl/environments/dataset.py:81]: Having a separate ChatEnv is useful, because given the interactive nature of this framework, we may end up needing a basic env that accumulates state in a convo between two agents for example, or interactively between agent and a user. It's also the simplest and most basic env to evaluate. We treat datasets as an extra delta on top where now the dataloader is providing one side of this data

File: forge/rl/environments/dataset.py [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+            batch_size=1,  # Always batch size 1 for episode loading
+            shuffle=self.shuffle,
+            generator=self.generator if self.shuffle else None,
+            collate_fn=self._identity_collate_fn,
+        )
+
+        # Initialize the iterator
+        self._data_iterator = iter(self.dataloader)
+
+        return self
```
Comment (ebsmothers) [inline forge/rl/environments/dataset.py:371]: Don't actually need to return anything here, right?

File: forge/rl/environments/dataset.py [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+
+        obs = env.reset()  # Loads initial conversation from dataset
+        # Continue conversation with additional turns via step()
+    """
+
+    def __init__(
+        self,
+        tokenizer: TokenizerProtocol,
+        dataset_name: Optional[str] = None,
+        dataset: Optional[Dataset] = None,
```
Comment (ebsmothers) [inline forge/rl/environments/dataset.py:126]: Kind of a small thing (and similar in spirit to a comment I left on the 1/3 PR): personally I prefer to just have a single blessed path for stuff like this rather than trying to support multiple paths via an if/else. It'll just keep the signature cleaner and the code easier to reason about imo

File: forge/rl/environments/dataset.py (hunk 1 above)
Comment (ebsmothers) [reply forge/rl/environments/dataset.py:81]: This is actually kinda related to Philip's comment (not to pile on too much here): I wonder if we should really inherit from ChatEnvironment at all. I know it's natural because all environments have the same methods. But really I think of the dataset chat environment as relying on ChatEnvironment as a subcomponent; fundamentally they are kind of different entities, no? Personally unless we are very clearly going to use ChatEnvironment as an environment (I'll defer to you on that, if you say we will I believe you), there is some cost to doing things like this -- we've now split the implementation in a way that's harder to understand and maintain (Btw if I am just being an anti-inheritance extremist please tell me 😃)

File: forge/rl/environments/dataset.py [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+        dataset_columns = set(self.dataset.column_names)
+        for mapping in self.column_mappings:
+            if mapping.column not in dataset_columns:
+                raise ValueError(
+                    f"Column '{mapping.column}' not found in dataset. Available columns: {', '.join(dataset_columns)}"
+                )
+
+        # Set up random seed for reproducibility
+        if seed is None:
+            seed = int(torch.empty((), dtype=torch.int64).random_().item())
```
Comment (ebsmothers) [inline forge/rl/environments/dataset.py:191]: In this case isn't it not reproducible though? Shouldn't we hardcode it? (Lmk if I'm missing the point here)

File: forge/rl/environments/dataset.py [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+
+        # Extract metadata and store in state
+        self._state.metadata = self._extract_metadata(row)
+
+        # Construct conversation from row
+        conversation = self._construct_conversation(row)
+
+        # Add all messages from the dataset to the conversation
+        for message in conversation:
+            action = self.message_to_action(message)
```
Comment (ebsmothers) [inline forge/rl/environments/dataset.py:270]: Related to my "no inheritance" comment: imo this is too much indirection for something that's basically just "tokenize the message"
